### ISSUE TITLE
Function fails with "Method not found" for DocumentClient constructor

### ISSUE THREAD

**Issue by JSkimming** [Azure/azure-functions-host#1679]
I've a C# Azure Functions project using the tooling in VS 2017 Preview (15.5.0 Preview 4.0)

I'm using the latest DocumentClient package [1.50.0](https://www.nuget.org/packages/Microsoft.Azure.DocumentDB/1.15.0).

At some point between version 1.11.4 and 1.15.0 (the code is not open sourced 👎) one of the constructors for DocumentClient change from:

```csharp
public DocumentClient(
    Uri serviceEndpoint,
    string authKeyOrResourceToken,
    ConnectionPolicy connectionPolicy = null,
    ConsistencyLevel? desiredConsistencyLevel = null)
```

to 

```csharp
public DocumentClient(
    Uri serviceEndpoint,
    string authKeyOrResourceToken,
    ConnectionPolicy connectionPolicy = null,
    ConsistencyLevel? desiredConsistencyLevel = null,
    JsonSerializerSettings serializerSettings = null) <-- Extra parameter
```

I've provided an [example repo](https://github.com/JSkimming/CosmosDbContructorError) the code change that caused the problem is this exact commit: https://github.com/JSkimming/CosmosDbContructorError/commit/47f7326e6fd2baff2437cc9dcb838c41e29c1555 (the upgrade to the Microsoft.Azure.DocumentDB package).

The problems occurs when running locally, and when deployed to Azure (see below).

It looks like the Functions run-time is referencing version [1.11.4 of the Microsoft.Azure.DocumentDB package](https://github.com/Azure/azure-webjobs-sdk-script/blob/v1.0.11027/src/WebJobs.Script.Host/packages.config#L14), whereas my project is referencing [1.15.0](https://github.com/JSkimming/CosmosDbContructorError/blob/master/CosmosDbError/CosmosDbError.csproj#L6). 1.11.4 seems to _win_ and because the constructor's signature changed the new version I'm referencing is not found at run-time.

#### Investigative information

- Timestamp: 2017-07-14T13:37:00.886 (and every minute thereafter)
- Function App name: CosmosDbError
- Function name: CreateClient
- Invocation ID: 55f7e6d5-0731-4dae-b696-fadd6c8df421
- Region: UK South

#### Repro steps

* Run this repo, the offending function (and only function) runs on a 1 minute timer.
  * https://github.com/JSkimming/CosmosDbContructorError
* You probably need the [Visual Studio 2017 Preview version 15.3](https://www.visualstudio.com/vs/preview/).

#### Expected behavior

No exception should be thrown.

#### Actual behavior

A Method not found exception is thrown, here's the logs from the function running in Azure:

```
2017-07-14T13:36:47  Welcome, you are now connected to log-streaming service.
2017-07-14T13:37:00.870 Function started (Id=55f7e6d5-0731-4dae-b696-fadd6c8df421)
2017-07-14T13:37:00.870 Function completed (Failure, Id=55f7e6d5-0731-4dae-b696-fadd6c8df421, Duration=0ms)
2017-07-14T13:37:00.886 Exception while executing function: Functions.CreateClient. mscorlib: Exception has been thrown by the target of an invocation. CosmosDbError: Method not found: 'Void Microsoft.Azure.Documents.Client.DocumentClient..ctor(System.Uri, System.String, Microsoft.Azure.Documents.Client.ConnectionPolicy, System.Nullable`1<Microsoft.Azure.Documents.ConsistencyLevel>, Newtonsoft.Json.JsonSerializerSettings)'.
```

#### Known workarounds

Downgrade to version 1.11.4 of the [Microsoft.Azure.DocumentDB NuGet](https://www.nuget.org/packages/Microsoft.Azure.DocumentDB/1.11.4) package

#### Related information 

* Programming language used : C#
* Links to source: https://github.com/JSkimming/CosmosDbContructorError/blob/master/CosmosDbError/CreateClient.cs#L19-L22

**Comment by JSkimming** [comment]
I've investigated further, and it looks like the constructor change came in between [1.14.1 -> 1.15.0](http://apicomparer.particular.net/compare/Microsoft.Azure.DocumentDB/1.14.1...1.15.0), therefore a downgrade to 1.14.1 resolved the constructor problem, though I suspect there is a wider issue.

One could argue that the Microsoft.Azure.DocumentDB package has broken [semantic versioning](http://semver.org/), by breaking binary compatibility within the same major version. SemVer does allow new APIs within the same major version, and this problem would occur if I were to reference a new API in the Microsoft.Azure.DocumentDB package.

And there are a lot of new APIs between 1.11.4 -> 1.15.0 (see below)

http://apicomparer.particular.net/compare/Microsoft.Azure.DocumentDB/1.15.0...1.11.4 (by reversing the numbers, new APIs are shown as removed).

**Comment by hershize** [comment]
Any update on this? I'm currently using DocumentDB version 1.15.0 as well.

**Comment by fabiocav** [comment]
The DocumentDB team is aware of the issue. The current workaround is to indeed downgrade until an update is made available.

**Comment by JSkimming** [comment]
@fabiocav While the DocumentDB SDK has broken binary compatibility, The problem would also occur should new APIs be released in the SDK. The SDK could legitimately be released under the same major version and conform to semantic versioning.

We would still suffer from _Method not found_ exceptions because the of the underling functions host is dependent on an earlier version of the SDK, that does not have the new APIs.

**Comment by fabiocav** [comment]
@JSkimming that is correct. We have plans to address some of those issues when using extension/binding types in function arguments (through a combination of tooling for the VS case, redirects when appropriate in the runtime and better errors). Some of those are mitigation steps to make the limitations clear and avoid running into those issues, but won't lift the requirement to match the version of the binding if you're using a type provided by the extension. If you are not, we do support side-by-side loading and will allow you to bind to a a different version.
A more permanent fix will be coming with the language out-of-proc and language extensibility work that will be landing in the next version.

**Comment by JSkimming** [comment]
@hershize It seems the breaking changes in the DocumentDB client have been fixed in [1.16.0](https://www.nuget.org/packages/Microsoft.Azure.DocumentDB/1.16.0). There are no breaking changes between [1.14.1 -> 1.16.0](http://apicomparer.particular.net/compare/Microsoft.Azure.DocumentDB/1.14.1...1.16.0).

If reverting the SDK is not an option, maybe upgrading works for you.

**Comment by hershize** [comment]
Thanks @JSkimming. Upgrading to 1.16.0 did solve the constructor problem. However, I'm running into other issues as well.

1. Querying by document ID doesn't correctly format the `Id` property as `id` in the expression.
2. Querying a document doesn't return the metadata of the document i.e. _ts, _etag, _attachments etc...

I just hope that the #992 short term solution will be able to solve this.

**Comment by kshyju** [comment]
Revisiting some older issues. This one appears to be quite dated, with the last update over seven years ago. I’m going to close it, as there is no immediate action needed at this time.

If you are currently encountering any issues with Azure Functions Cosmos DB bindings, please feel free to open a new GitHub issue with relevant details. We will be happy to take a closer look.
